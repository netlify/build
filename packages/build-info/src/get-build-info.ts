import { listFrameworks } from '@netlify/framework-info'

import { ContextOptions, getContext } from './context.js'
import { BuildSystem, detectBuildSystems } from './detect-build-system.js'
import { detectPackageManager, PkgManagerFields } from './detect-package-manager.js'
import { getWorkspaceInfo, WorkspaceInfo } from './workspaces.js'

export type Info = {
  jsWorkspaces?: WorkspaceInfo
  packageManager?: PkgManagerFields
  frameworks: unknown[]
  buildSystems?: BuildSystem[]
}

export const getBuildInfo = async (opts: ContextOptions) => {
  const context = await getContext(opts)
  let frameworks: any[] = []

  try {
    // if the framework  detection is crashing we should not crash the build info and package-manager
    // detection
    frameworks = await listFrameworks({ projectDir: context.projectDir })
  } catch {
    // TODO: build reporting to buildbot see: https://github.com/netlify/pillar-workflow/issues/1001
    // noop
  }

  const info: Info = { frameworks }

  try {
    info.buildSystems = await detectBuildSystems(context.projectDir, context.rootDir)
  } catch (error) {
    // noop
  }

  // only if we find a root package.json we know this is a javascript workspace
  if (Object.keys(context.rootPackageJson).length > 0) {
    info.packageManager = await detectPackageManager(context.projectDir, context.rootDir)
    const workspaceInfo = await getWorkspaceInfo(info.packageManager, context)
    if (workspaceInfo) {
      info.jsWorkspaces = workspaceInfo
    }
  }

  return info
}
