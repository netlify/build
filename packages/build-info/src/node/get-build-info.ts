import { listFrameworks } from '@netlify/framework-info'

import { Logger } from '../file-system.js'
import { report } from '../metrics.js'
import { PkgManagerFields } from '../package-managers/detect-package-manager.js'
import { Project } from '../project.js'
import { WorkspaceInfo } from '../workspaces/detect-workspace.js'

import { NodeFS } from './file-system.js'

export type Info = {
  jsWorkspaces: WorkspaceInfo | null
  packageManager: PkgManagerFields | null
  frameworks: unknown[]
  buildSystems: {
    name: string
    version?: string | undefined
  }[]
}

/** A noop logger that is used to not log anything (we use the stdout for parsing the json output) */
export class NoopLogger implements Logger {
  debug() {
    /** noop */
  }
  log() {
    /** noop */
  }
  error() {
    /** noop */
  }
  info() {
    /** noop */
  }
  warn() {
    /** noop */
  }
}

/** Get the build info object that is used inside buildbot */
export async function getBuildInfo(
  config: {
    projectDir?: string
    rootDir?: string
    featureFlags?: Record<string, boolean>
  } = { featureFlags: {} },
): Promise<Info> {
  const fs = new NodeFS()
  // prevent logging in output as we use the stdout to capture the json
  fs.logger = new NoopLogger()
  const project = new Project(fs, config.projectDir, config.rootDir)
    .setEnvironment(process.env)
    .setNodeVersion(process.version)

  const info: Info = {
    packageManager: await project.detectPackageManager(),
    jsWorkspaces: await project.detectWorkspaces(),
    frameworks: [],
    buildSystems: await project.detectBuildSystem(),
  }

  if (config.featureFlags?.build_info_new_framework_detection) {
    info.frameworks = (await project.detectFrameworksInPath(project.baseDirectory)) || []
  } else {
    try {
      // if the framework  detection is crashing we should not crash the build info and package-manager detection
      info.frameworks = await listFrameworks({ projectDir: project.baseDirectory })
    } catch (error) {
      report(error)
    }
  }

  return info
}
