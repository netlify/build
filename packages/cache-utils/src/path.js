'use strict'

const { homedir } = require('os')
const { resolve, isAbsolute, join, sep } = require('path')

const { getCacheDir } = require('./dir')
const { safeGetCwd } = require('./utils/cwd')

// Find the paths of the file before/after caching
const parsePath = async function ({ path, cacheDir, cwdOpt, mode }) {
  const srcPath = await getSrcPath(path, cwdOpt)
  const cachePath = await getCachePath({ srcPath, cacheDir, cwdOpt, mode })
  return { srcPath, cachePath }
}

// Retrieve absolute path to the local file to cache/restore
const getSrcPath = async function (path, cwdOpt) {
  const cwd = await safeGetCwd(cwdOpt)
  const srcPath = resolvePath(path, cwd)
  checkSrcPath(srcPath, cwd)
  return srcPath
}

const resolvePath = function (path, cwd) {
  if (isAbsolute(path)) {
    return resolve(path)
  }

  if (cwd !== '') {
    return resolve(cwd, path)
  }

  throw new Error(`Current directory does not exist: ${cwd}`)
}

// Caching the whole repository creates many issues:
//  - It caches many directories that are not related to Gatsby but take lots of
//    space, such as node_modules
//  - It caches directories that are not meant to restored across builds. For
//    example .git (beside being big).
//  - It undoes any build operations inside the repository that might have
//    happened before this plugin starts restoring the cache, leading to
//    conflicts with other plugins, Netlify Build or the buildbot.
const checkSrcPath = function (srcPath, cwd) {
  if (cwd !== '' && isParentPath(srcPath, cwd)) {
    throw new Error(`Cannot cache ${srcPath} because it is the current directory (${cwd}) or a parent directory`)
  }
}

// Note: srcPath and cwd are already normalized and absolute
const isParentPath = function (srcPath, cwd) {
  return `${cwd}${sep}`.startsWith(`${srcPath}${sep}`)
}

const getCachePath = async function ({ srcPath, cacheDir, cwdOpt, mode }) {
  const cacheDirA = await getCacheDir({ cacheDir, mode })
  const { name, relPath } = await findBase(srcPath, cwdOpt)
  const cachePath = join(cacheDirA, name, relPath)
  return cachePath
}

// The cached path is the path relative to the base which can be either the
// current directory, the home directory or the root directory. Each is tried
// in order.
const findBase = async function (srcPath, cwdOpt) {
  const bases = await getBases(cwdOpt)
  const srcPathA = normalizeWindows(srcPath)
  return bases.map(({ name, base }) => parseBase(name, base, srcPathA)).find(Boolean)
}

// Windows drives are problematic:
//  - they cannot be used in `relPath` since directories cannot be called `C:`
//    for example.
//  - they make comparing between drives harder
// For the moment, we simply strip them. This does mean two files with the same
// paths but inside different drives would be cached together, which is not
// correct.
// This also means `cache.list()` always assumes the source files are on the
// current drive.
// TODO: fix it.
const normalizeWindows = function (srcPath) {
  return srcPath.replace(WINDOWS_DRIVE_REGEX, '\\')
}

const WINDOWS_DRIVE_REGEX = /^[a-zA-Z]:\\/

// This logic works when `base` and `path` are on different Windows drives
const parseBase = function (name, base, srcPath) {
  if (srcPath === base || !srcPath.startsWith(base)) {
    return
  }

  const relPath = srcPath.replace(base, '')
  return { name, relPath }
}

const getBases = async function (cwdOpt) {
  const cwdBase = await getCwdBase(cwdOpt)
  return [...cwdBase, { name: 'home', base: homedir() }, { name: 'root', base: sep }]
}

const getCwdBase = async function (cwdOpt) {
  const cwd = await safeGetCwd(cwdOpt)
  if (cwd === '') {
    return []
  }

  return [{ name: 'cwd', base: cwd }]
}

module.exports = { parsePath, getBases }
