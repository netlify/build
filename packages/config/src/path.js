const { resolve } = require('path')

const findUp = require('find-up')
const pLocate = require('p-locate')
const pathExists = require('path-exists')

const { resolvePath } = require('./files')

// Configuration location can be:
//  - a local path with the --config CLI flag
//  - a `netlify.*` file in the `repositoryRoot/{base}`
//  - a `netlify.*` file in the `repositoryRoot`
//  - a `netlify.*` file in the current directory or any parent
const getConfigPath = async function({ configOpt, cwd, repositoryRoot, base }) {
  const configPath = await pLocate(
    [
      searchConfigOpt(cwd, configOpt),
      searchBaseConfigFile(repositoryRoot, base),
      searchConfigFile(repositoryRoot),
      findUp(FILENAME, { cwd }),
    ],
    Boolean,
  )
  return configPath
}

// --config CLI flag
const searchConfigOpt = function(cwd, configOpt) {
  if (configOpt === undefined) {
    return
  }

  return resolve(cwd, configOpt)
}

// Look for `repositoryRoot/{base}/netlify.*`
const searchBaseConfigFile = function(repositoryRoot, base) {
  if (base === undefined) {
    return
  }

  const basePath = resolvePath(repositoryRoot, base)
  return searchConfigFile(basePath)
}

// Look for several file extensions for `netlify.*`
const searchConfigFile = async function(cwd) {
  const path = resolve(cwd, FILENAME)
  if (!(await pathExists(path))) {
    return
  }
  return path
}

const FILENAME = 'netlify.toml'

module.exports = { getConfigPath }
