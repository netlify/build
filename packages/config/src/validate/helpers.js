'use strict'

const isPlainObj = require('is-plain-obj')

const isArrayOfObjects = function (value) {
  return Array.isArray(value) && value.every(isPlainObj)
}

const isArrayOfStrings = function (value) {
  return Array.isArray(value) && value.every(isString)
}

const isString = function (value) {
  return typeof value === 'string'
}

// Check an object valid properties, including legacy ones
const validProperties = function (propNames, legacyPropNames) {
  return {
    check: (value) => checkValidProperty(value, [...propNames, ...legacyPropNames]),
    message: `has unknown properties. Valid properties are:
${propNames.map((propName) => `  - ${propName}`).join('\n')}`,
  }
}

const checkValidProperty = function (value, propNames) {
  return Object.keys(value).every((propName) => propNames.includes(propName))
}

const functionsDirectoryCheck = {
  formatInvalid: ({ functionsDirectory } = {}) => ({ functions: { directory: functionsDirectory } }),
  propertyName: 'functions.directory',
}

module.exports = {
  isArrayOfObjects,
  isArrayOfStrings,
  isString,
  validProperties,
  functionsDirectoryCheck,
}
