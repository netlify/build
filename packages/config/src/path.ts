import { existsSync } from 'fs'
import { resolve } from 'path'

import { findUp } from 'find-up'
import pLocate from 'p-locate'

/**
 * Configuration location can be:
 * - a local path with the --config CLI flag
 * - a `netlify.*` file in the `repositoryRoot/{base}`
 * - a `netlify.*` file in the `repositoryRoot`
 * - a `netlify.*` file in the current directory or any parent
 */
export const getConfigPath = async function ({ configOpt, cwd, repositoryRoot, configBase }) {
  const configPath = await pLocate(
    [
      searchConfigOpt(cwd, configOpt),
      searchBaseConfigFile(configBase),
      searchConfigFile(repositoryRoot),
      findUp(FILENAME, { cwd }),
    ],
    Boolean,
  )
  return configPath
}

/** --config CLI flag */
const searchConfigOpt = function (cwd: string, configOpt?: string) {
  if (configOpt === undefined || configOpt.length === 0) {
    return
  }

  return resolve(cwd, configOpt)
}

/**
 * Look for `repositoryRoot/{base}/netlify.*`
 */
const searchBaseConfigFile = function (configBase?: string) {
  if (configBase === undefined) {
    return
  }

  return searchConfigFile(configBase)
}

/**
 * Look for several file extensions for `netlify.*`
 */
const searchConfigFile = function (cwd: string): string | undefined {
  const path = resolve(cwd, FILENAME)
  if (!existsSync(path)) {
    return
  }
  return path
}

const FILENAME = 'netlify.toml'
