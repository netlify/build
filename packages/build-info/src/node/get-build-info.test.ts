import { platform } from 'os'
import { join } from 'path'

import { afterEach, describe, expect, test } from 'vitest'

import { createFixture } from '../../tests/helpers.js'
import { mockFileSystem } from '../../tests/mock-file-system.js'
import { getBuildInfo } from '../node/get-build-info.js'

afterEach(async ({ cleanup }) => await cleanup?.())

test('should detect nothing in an empty project', async () => {
  const cwd = mockFileSystem({})
  const info = await getBuildInfo({ projectDir: cwd })
  expect(info).toMatchInlineSnapshot(`
    {
      "buildSystems": [],
      "frameworks": [],
      "jsWorkspaces": null,
      "langRuntimes": [],
      "packageManager": null,
      "settings": [],
    }
  `)
})

test('should only detect go in a simple golang project', async () => {
  const cwd = mockFileSystem({
    'go.mod': '',
    'main.go': '',
  })
  const info = await getBuildInfo({ projectDir: cwd })
  expect(info).toMatchInlineSnapshot(`
    {
      "buildSystems": [],
      "frameworks": [],
      "jsWorkspaces": null,
      "langRuntimes": [
        {
          "id": "go",
          "name": "Go",
        },
      ],
      "packageManager": null,
      "settings": [],
    }
  `)
})

test('should not crash on invalid projects', async (ctx) => {
  const fixture = await createFixture('invalid-project', ctx)
  const { frameworks, packageManager } = await getBuildInfo({ projectDir: fixture.cwd })
  expect(packageManager).toMatchInlineSnapshot
  expect(frameworks).toEqual([])
})

test.skipIf(platform() === 'win32')('should retrieve the build info for providing a rootDir', async (ctx) => {
  const fixture = await createFixture('pnpm-workspace', ctx)
  const info = await getBuildInfo({ rootDir: fixture.cwd })
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  info.jsWorkspaces!.rootDir = '/cleaned-for-snapshot'
  expect(info).toMatchSnapshot()
})

test.skipIf(platform() === 'win32')(
  'should retrieve the build info for providing a rootDir and the same projectDir',
  async (ctx) => {
    const fixture = await createFixture('pnpm-workspace', ctx)
    const info = await getBuildInfo({ rootDir: fixture.cwd, projectDir: fixture.cwd })
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    info.jsWorkspaces!.rootDir = '/cleaned-for-snapshot'
    expect(info).toMatchSnapshot()
  },
)

test.skipIf(platform() === 'win32')(
  'should retrieve the build info for providing a rootDir and a nested projectDir',
  async (ctx) => {
    const fixture = await createFixture('pnpm-workspace', ctx)
    const info = await getBuildInfo({ rootDir: fixture.cwd, projectDir: join(fixture.cwd, 'packages/blog') })
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    info.jsWorkspaces!.rootDir = '/cleaned-for-snapshot'
    expect(info).toMatchSnapshot()
  },
)

describe('Golang', () => {
  test('should only detect Go inside a golang workspace', async (ctx) => {
    const fixture = await createFixture('go-workspace', ctx)
    const info = await getBuildInfo({ projectDir: 'bar', rootDir: fixture.cwd })
    expect(info).toMatchInlineSnapshot(`
      {
        "buildSystems": [],
        "frameworks": [],
        "jsWorkspaces": null,
        "langRuntimes": [
          {
            "id": "go",
            "name": "Go",
          },
        ],
        "packageManager": null,
        "settings": [],
      }
    `)
  })
})

describe('JavaScript Workspaces', () => {
  test('project without package.json does not return workspaces info', async (ctx) => {
    const fixture = await createFixture('empty', ctx)
    const { jsWorkspaces } = await getBuildInfo({ projectDir: fixture.cwd })

    expect(jsWorkspaces).toBeNull()
  })

  test('project without workspaces in package.json does not return workspaces info', async (ctx) => {
    const fixture = await createFixture('simple-package-json', ctx)
    const { jsWorkspaces } = await getBuildInfo({ projectDir: fixture.cwd })
    expect(jsWorkspaces).toBeNull()
  })

  test('projectDir set to workspaces root returns workspace info and isRoot flag set to true', async (ctx) => {
    const fixture = await createFixture('js-workspaces', ctx)
    const { jsWorkspaces } = await getBuildInfo({ projectDir: fixture.cwd })
    expect(jsWorkspaces).not.toBeNull()
    expect(jsWorkspaces?.isRoot).toBe(true)
    expect(jsWorkspaces?.packages.length).toBe(2)
  })

  test('projectDir set to workspace dir returns workspace info and isRoot flag set to false', async (ctx) => {
    const fixture = await createFixture('js-workspaces', ctx)
    const { jsWorkspaces } = await getBuildInfo({ projectDir: 'packages/gatsby-site', rootDir: fixture.cwd })
    expect(jsWorkspaces).not.toBeNull()
    expect(jsWorkspaces?.isRoot).toBe(false)
    expect(jsWorkspaces?.packages.length).toBe(2)
  })

  test('if project is not part of a workspace return no workspace info', async (ctx) => {
    const fixture = await createFixture('js-workspaces', ctx)
    const { jsWorkspaces } = await getBuildInfo({ projectDir: 'not-in-workspace', rootDir: fixture.cwd })
    expect(jsWorkspaces).toBeNull()
  })

  test('handles absolute paths correctly', async (ctx) => {
    const fixture = await createFixture('js-workspaces', ctx)
    const { jsWorkspaces } = await getBuildInfo({
      projectDir: `${fixture.cwd}/packages/gatsby-site`,
      rootDir: fixture.cwd,
    })
    expect(jsWorkspaces).not.toBeNull()
    expect(jsWorkspaces?.isRoot).toBe(false)
    expect(jsWorkspaces?.packages.length).toBe(2)
  })

  test('detects workspaces and frameworks when given a rootDir and an empty projectDir', async (ctx) => {
    const fixture = await createFixture('js-workspaces', ctx)
    const { frameworks, jsWorkspaces } = await getBuildInfo({ projectDir: '', rootDir: fixture.cwd })
    expect(jsWorkspaces).not.toBeNull()
    expect(frameworks.length).toBe(1)
  })

  test('should detect workspaces and frameworks', async (ctx) => {
    const fixture = await createFixture('js-workspaces', ctx)
    const { frameworks, jsWorkspaces } = await getBuildInfo({
      projectDir: 'packages/gatsby-site',
      rootDir: fixture.cwd,
    })
    expect(jsWorkspaces).not.toBeNull()
    expect(frameworks.length).toBe(1)
  })
})

describe('Frameworks', () => {
  test('return an empty array if no frameworks are detected', async (ctx) => {
    const fixture = await createFixture('empty', ctx)
    const { frameworks } = await getBuildInfo({ projectDir: fixture.cwd })
    expect(frameworks).toEqual([])
  })

  test('framework detection works correctly for npm pkg detected framework', async (ctx) => {
    const fixture = await createFixture('next-project', ctx)
    const { frameworks } = await getBuildInfo({ projectDir: fixture.cwd })
    expect(frameworks).toHaveLength(1)
    expect(frameworks).toEqual([expect.objectContaining({ id: 'next' })])
  })

  test('framework detection works correctly for static file detected framework', async (ctx) => {
    const fixture = await createFixture('jekyll-project', ctx)
    const { frameworks } = await getBuildInfo({ projectDir: fixture.cwd })
    expect(frameworks).toHaveLength(1)
    expect(frameworks).toEqual([expect.objectContaining({ id: 'jekyll' })])
  })
})
