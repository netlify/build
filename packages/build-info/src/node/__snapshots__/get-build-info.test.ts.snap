// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should retrieve the build info for providing a rootDir 1`] = `
{
  "buildSystems": [
    {
      "id": "pnpm",
      "logo": undefined,
      "name": "PNPM",
      "version": undefined,
    },
  ],
  "frameworks": [],
  "jsWorkspaces": WorkspaceInfo {
    "isRoot": true,
    "packages": [
      {
        "name": "@evilcorp/blog",
        "path": "packages/blog",
      },
      {
        "name": "@evilcorp/website",
        "path": "packages/website",
      },
    ],
    "rootDir": "/cleaned-for-snapshot",
  },
  "langRuntimes": [
    {
      "id": "node",
      "name": "NodeJS",
    },
  ],
  "packageManager": {
    "forceEnvironment": "NETLIFY_USE_PNPM",
    "installCommand": "pnpm install",
    "lockFile": "pnpm-lock.yaml",
    "name": "pnpm",
    "runCommand": "pnpm run",
    "version": SemVer {
      "build": [],
      "includePrerelease": false,
      "loose": false,
      "major": 6,
      "minor": 32,
      "options": {},
      "patch": 18,
      "prerelease": [],
      "raw": "6.32.18",
      "version": "6.32.18",
    },
  },
  "settings": [
    {
      "baseDirectory": "",
      "buildCommand": "pnpm --filter @evilcorp/blog run build",
      "devCommand": "pnpm --filter @evilcorp/blog run dev",
      "dist": "packages/blog/dist",
      "env": {},
      "framework": {
        "id": "astro",
        "name": "Astro",
      },
      "frameworkPort": 3000,
      "name": "PNPM + Astro packages/blog",
      "packagePath": "packages/blog",
      "plugins_from_config_file": [],
      "plugins_recommended": [],
      "pollingStrategies": [
        "TCP",
        "HTTP",
      ],
    },
    {
      "baseDirectory": "",
      "buildCommand": "pnpm --filter @evilcorp/website run build",
      "devCommand": "pnpm --filter @evilcorp/website run dev",
      "dist": "packages/website/.next",
      "env": {},
      "framework": {
        "id": "next",
        "name": "Next.js",
      },
      "frameworkPort": 3000,
      "name": "PNPM + Next.js packages/website",
      "packagePath": "packages/website",
      "plugins_from_config_file": [],
      "plugins_recommended": [
        "@netlify/plugin-nextjs",
      ],
      "pollingStrategies": [
        "TCP",
      ],
    },
  ],
}
`;

exports[`should retrieve the build info for providing a rootDir and a nested projectDir 1`] = `
{
  "buildSystems": [
    {
      "id": "pnpm",
      "logo": undefined,
      "name": "PNPM",
      "version": undefined,
    },
  ],
  "frameworks": [],
  "jsWorkspaces": WorkspaceInfo {
    "isRoot": false,
    "packages": [
      {
        "name": "@evilcorp/blog",
        "path": "packages/blog",
      },
      {
        "name": "@evilcorp/website",
        "path": "packages/website",
      },
    ],
    "rootDir": "/cleaned-for-snapshot",
  },
  "langRuntimes": [
    {
      "id": "node",
      "name": "NodeJS",
    },
  ],
  "packageManager": {
    "forceEnvironment": "NETLIFY_USE_PNPM",
    "installCommand": "pnpm install",
    "lockFile": "pnpm-lock.yaml",
    "name": "pnpm",
    "runCommand": "pnpm run",
    "version": SemVer {
      "build": [],
      "includePrerelease": false,
      "loose": false,
      "major": 6,
      "minor": 32,
      "options": {},
      "patch": 18,
      "prerelease": [],
      "raw": "6.32.18",
      "version": "6.32.18",
    },
  },
  "settings": [
    {
      "baseDirectory": "",
      "buildCommand": "pnpm --filter @evilcorp/blog run build",
      "devCommand": "pnpm --filter @evilcorp/blog run dev",
      "dist": "packages/blog/dist",
      "env": {},
      "framework": {
        "id": "astro",
        "name": "Astro",
      },
      "frameworkPort": 3000,
      "name": "PNPM + Astro packages/blog",
      "packagePath": "packages/blog",
      "plugins_from_config_file": [],
      "plugins_recommended": [],
      "pollingStrategies": [
        "TCP",
        "HTTP",
      ],
    },
  ],
}
`;

exports[`should retrieve the build info for providing a rootDir and the same projectDir 1`] = `
{
  "buildSystems": [
    {
      "id": "pnpm",
      "logo": undefined,
      "name": "PNPM",
      "version": undefined,
    },
  ],
  "frameworks": [],
  "jsWorkspaces": WorkspaceInfo {
    "isRoot": true,
    "packages": [
      {
        "name": "@evilcorp/blog",
        "path": "packages/blog",
      },
      {
        "name": "@evilcorp/website",
        "path": "packages/website",
      },
    ],
    "rootDir": "/cleaned-for-snapshot",
  },
  "langRuntimes": [
    {
      "id": "node",
      "name": "NodeJS",
    },
  ],
  "packageManager": {
    "forceEnvironment": "NETLIFY_USE_PNPM",
    "installCommand": "pnpm install",
    "lockFile": "pnpm-lock.yaml",
    "name": "pnpm",
    "runCommand": "pnpm run",
    "version": SemVer {
      "build": [],
      "includePrerelease": false,
      "loose": false,
      "major": 6,
      "minor": 32,
      "options": {},
      "patch": 18,
      "prerelease": [],
      "raw": "6.32.18",
      "version": "6.32.18",
    },
  },
  "settings": [
    {
      "baseDirectory": "",
      "buildCommand": "pnpm --filter @evilcorp/blog run build",
      "devCommand": "pnpm --filter @evilcorp/blog run dev",
      "dist": "packages/blog/dist",
      "env": {},
      "framework": {
        "id": "astro",
        "name": "Astro",
      },
      "frameworkPort": 3000,
      "name": "PNPM + Astro packages/blog",
      "packagePath": "packages/blog",
      "plugins_from_config_file": [],
      "plugins_recommended": [],
      "pollingStrategies": [
        "TCP",
        "HTTP",
      ],
    },
    {
      "baseDirectory": "",
      "buildCommand": "pnpm --filter @evilcorp/website run build",
      "devCommand": "pnpm --filter @evilcorp/website run dev",
      "dist": "packages/website/.next",
      "env": {},
      "framework": {
        "id": "next",
        "name": "Next.js",
      },
      "frameworkPort": 3000,
      "name": "PNPM + Next.js packages/website",
      "packagePath": "packages/website",
      "plugins_from_config_file": [],
      "plugins_recommended": [
        "@netlify/plugin-nextjs",
      ],
      "pollingStrategies": [
        "TCP",
      ],
    },
  ],
}
`;
